### PR TITLE
[AIRFLOW-1140] DatabricksSubmitRunOperator should template the "json" field

### PR DESCRIPTION
**Pull request by andrewmchen**
Dear Airflow maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "[AIRFLOW-XXX] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-XXX

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:

For the DatabricksSubmitRunOperator, we'd like to have the JSON parameter be templated.

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: airflow/contrib/operators/databricks_operator.py [hunk 1]
```diff
@@ -120,6 +129,7 @@ class DatabricksSubmitRunOperator(BaseOperator):
         unreachable. Its value must be greater than or equal to 1.
     :type databricks_retry_limit: int
     """
+    template_fields = ('json',)
```
Comment (mengxr) [inline airflow/contrib/operators/databricks_operator.py]: Could you comment inline which class uses this field?

File: airflow/contrib/operators/databricks_operator.py [hunk 2]
```diff
@@ -81,34 +82,42 @@ class DatabricksSubmitRunOperator(BaseOperator):
         (i.e. ``spark_jar_task``, ``notebook_task``..) to this operator will
         be merged with this json dictionary if they are provided.
         If there are conflicts during the merge, the named parameters will
-        take precedence and override the top level json keys.
+        take precedence and override the top level json keys. This field will be
+        templated.
```
Comment (mengxr) [inline airflow/contrib/operators/databricks_operator.py]: Please add a `seealso` link to for templates in Airflow.

File: airflow/contrib/operators/databricks_operator.py [hunk 3]
```diff
@@ -163,9 +173,31 @@ def __init__(
         if 'run_name' not in self.json:
             self.json['run_name'] = run_name or kwargs['task_id']
 
+        self.json = self._deep_string_coerce(self.json)
         # This variable will be used in case our task gets killed.
         self.run_id = None
 
+    def _deep_string_coerce(self, content, json_path='json'):
+        """
+        The ``self.json`` field must be a dict with only string values. This is because
```
Comment (mengxr) [inline airflow/contrib/operators/databricks_operator.py]: Document what this method does first before explaining why we need it.

File: airflow/contrib/operators/databricks_operator.py [hunk 4]
```diff
@@ -163,9 +173,31 @@ def __init__(
...
+        ``render_template`` will fail for numerical values.
+        """
+        c = self._deep_string_coerce
+        if isinstance(content, six.string_types):
+            return content
+        elif isinstance(content, six.integer_types+(float,)):
+            return str(content)
```
Comment (mengxr) [inline airflow/contrib/operators/databricks_operator.py]: Does converting numbers into strings introduce any type issues? Probably need an inline comment to assure readers that this is safe.

File: airflow/contrib/operators/databricks_operator.py [hunk 5]
```diff
@@ -163,9 +173,31 @@ def __init__(
...
         # This variable will be used in case our task gets killed.
         self.run_id = None
 
+    def _deep_string_coerce(self, content, json_path='json'):
+        """
+        The ``self.json`` field must be a dict with only string values. This is because
+        ``render_template`` will fail for numerical values.
+        """
+        c = self._deep_string_coerce
+        if isinstance(content, six.string_types):
```
Comment (mengxr) [inline airflow/contrib/operators/databricks_operator.py]: check `content is None`

**Review by mengxr — CHANGES_REQUESTED** [review]

**Comment by mengxr** [comment]
LGTM. @artwr Could you make a pass?

**Comment by artwr** [comment]
Looks like it was merged but LGTM +1 for posterity.

**Pull request closed without merging**